connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ExecuteResponseObserver.scala

```diff
@@ -154,11 +206,27 @@ private[connect] class ExecuteResponseObserver[T](val executeHolder: ExecuteHold
   /**
    * Remove cached responses after response with lastReturnedIndex is returned from getResponse.
    * Remove according to caching policy:
-   *   - if query is not reattachable, remove all responses up to and including
-   *     highestConsumedIndex.
+   *   - if retryBufferSize is 0 (or query is not reattachable), remove all responses up to and
+   *     including lastSentIndex.
+   *   - otherwise keep responses backwards from lastSentIndex until their total size exceeds
+   *     retryBufferSize
    */
-  private def removeCachedResponses() = {
-    var i = highestConsumedIndex
+  private def removeCachedResponses(lastSentIndex: Long) = {
+    var i = lastSentIndex
+    var totalResponsesSize = 0L
+    while (i >= 1 && responses.get(i).isDefined && totalResponsesSize < retryBufferSize) {
+      totalResponsesSize += responses.get(i).get.serializedByteSize
```

Comment:
Is the total response size only update when we remove the items?

Reply:
This is indeed calculated on the fly when removed, and iterates back from the returned response until retryBufferSize is kept. This could be smarter, we could keep some "second finger" on the oldest buffered, and then cache the total between that and the highest consumed (our "first finger"), but it does actually get a bit more complicated when after a reattach the client backtracks (so our "first finger" moves back)